### DISCUSSION TITLE
How to log HTTP request message that is / would be sent to odata endpoint

### DISCUSSION THREAD

**Discussion by jwevansgam** [OData/odata.net#2568] [Q&A]
I am trying to diagnose an issue when calling a customers odata api - the service client has been created using the Visual Studio add service connection and is using Microsoft.OData.Client 7.10.0. The call is using a being / end pattern and calling GetValueAsync on the service client. The call is succeeding but I am getting an error back from the customers api and would like to send them the request that I sending so they are able to diagnose on their side.

Does anyone know how I can write out the HTTP request message before it is sent to the customer api?

I have tried fidder but this doesn't work in our corporate environment and I have also tried intercepting the request using SendingRequest2 but the request doesn't seem to be available when calling GetStream.

I am running the project locally so I can make any code changes necessary to just capture the output message - it doesn't need to be part of an operational system so not looking for perfect solutions - even just logging out to console would do!

**Comment by habbes** [comment]
You could hook into the client request events (https://learn.microsoft.com/en-us/odata/client/using-hooks), specifically the [`SendingRequest2` event](https://learn.microsoft.com/en-us/dotnet/api/microsoft.odata.client.dataservicecontext.sendingrequest2?view=odata-client-7.0). The event args contain a `RequestMessage` property that you can use to inspect the request headers and stream.
